File: grove-web/src/api/sketches.ts
// Sketches API client
// Backend routes are mounted under /api/v1/projects/{id}/tasks/{taskId}/sketches

import { apiClient, getApiHost } from './client';

export interface SketchMeta {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
}

export interface SketchListResponse {
  sketches: SketchMeta[];
}

export async function listSketches(
  projectId: string,
  taskId: string,
): Promise<SketchMeta[]> {
  const data = await apiClient.get<SketchListResponse>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches`,
  );
  return data.sketches;
}

export async function createSketch(
  projectId: string,
  taskId: string,
  name: string,
): Promise<SketchMeta> {
  return apiClient.post<{ name: string }, SketchMeta>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches`,
    { name },
  );
}

export async function deleteSketch(
  projectId: string,
  taskId: string,
  sketchId: string,
): Promise<void> {
  await apiClient.delete(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}`,
  );
}

export async function renameSketch(
  projectId: string,
  taskId: string,
  sketchId: string,
  name: string,
): Promise<void> {
  await apiClient.post<{ name: string }, unknown>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}/rename`,
    { name },
  );
}

export async function getSketchScene(
  projectId: string,
  taskId: string,
  sketchId: string,
): Promise<unknown> {
  return apiClient.get<unknown>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}`,
  );
}

export async function putSketchScene(
  projectId: string,
  taskId: string,
  sketchId: string,
  scene: unknown,
): Promise<void> {
  await apiClient.put<{ scene: unknown }, unknown>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}`,
    { scene },
  );
}

/**
 * Flush variant used from `beforeunload` / last-chance flush paths. Uses
 * `fetch(..., { keepalive: true })` so the request survives page unload.
 * Browser-capped at ~64 KB body; larger scenes silently fall back by
 * throwing — callers should handle that and drop to a best-effort regular
 * PUT (which may itself be aborted by the browser on unload).
 */
export async function putSketchSceneKeepalive(
  projectId: string,
  taskId: string,
  sketchId: string,
  scene: unknown,
): Promise<void> {
  await apiClient.putKeepalive<{ scene: unknown }>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}`,
    { scene },
  );
}

/**
 * Upload a PNG thumbnail for a sketch. The server writes it unconditionally;
 * staleness is detected at MCP-read time by comparing file mtimes (thumb vs
 * scene), so a thumbnail rendered slightly before a new scene write is simply
 * ignored on read instead of being rejected on write.
 */
export async function uploadSketchThumbnail(
  projectId: string,
  taskId: string,
  sketchId: string,
  png: Blob,
): Promise<void> {
  const path = `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}/thumbnail`;
  await apiClient.postBinary(path, png, 'image/png');
}

export interface SketchHistoryEntry {
  id: string;
  ts: string;
  element_count?: number | null;
  label?: string | null;
}

export interface SketchHistoryResponse {
  entries: SketchHistoryEntry[];
}

/**
 * List checkpoints available to restore for this sketch. Checkpoints live
 * per-sketch — the returned list is already scoped to `sketchId`.
 */
export async function listSketchHistory(
  projectId: string,
  taskId: string,
  sketchId: string,
): Promise<SketchHistoryEntry[]> {
  const data = await apiClient.get<SketchHistoryResponse>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}/history`,
  );
  return data.entries;
}

/**
 * Restore the given checkpoint onto this sketch, overwriting its current
 * scene. Server broadcasts a `sketch_updated` event (agent source) so
 * connected clients hard-reload the canvas.
 */
export async function restoreSketchCheckpoint(
  projectId: string,
  taskId: string,
  sketchId: string,
  checkpointId: string,
): Promise<void> {
  await apiClient.post<{ checkpoint_id: string }, unknown>(
    `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}/restore`,
    { checkpoint_id: checkpointId },
  );
}

/**
 * Build the URL for the sketch's PNG render. Returned synchronously so it can
 * be dropped straight into an `<img src>` (HMAC for mobile is handled by the
 * server's per-host policy, mirroring the existing `<img>` markdown path).
 * The endpoint returns 404 when no thumbnail has been uploaded yet — callers
 * should fall back to plain text on `onError`.
 */
export function sketchThumbnailUrl(
  projectId: string,
  taskId: string,
  sketchId: string,
): string {
  return `/api/v1/projects/${projectId}/tasks/${taskId}/sketches/${sketchId}/thumbnail`;
}

/**
 * Build a WebSocket URL for the sketches live-update endpoint.
 * The caller is responsible for appending HMAC auth (see `appendHmacToUrl`).
 */
export function sketchWsUrl(projectId: string, taskId: string): string {
  const host = getApiHost();
  const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
  return `${proto}//${host}/api/v1/projects/${projectId}/tasks/${taskId}/sketches/ws`;
}

File: src/api/handlers/tasks/sketches.rs
//! REST handlers for Studio task sketches.

use axum::{
    body::Bytes,
    extract::Path,
    http::{header, HeaderMap, HeaderValue, StatusCode},
    response::{IntoResponse, Response},
    Json,
};
use serde::{Deserialize, Serialize};

use crate::error::GroveError;
use crate::storage::sketches;

use super::super::common::find_project_by_id;
use super::sketch_events::{broadcast_sketch_event, SketchEvent, SketchEventSource};

/// Map a storage error from a scene write / patch into an HTTP response that
/// preserves the detailed message. `GroveError::Storage(_)` values are all
/// produced by our own validation paths (bad sketch id, missing/duplicate
/// element id, non-object update, empty-string delete, id conflict, scene
/// missing elements) and are client bugs → 400 with the message so the AI
/// agent / UI can self-correct. Everything else (I/O, JSON serde on disk
/// contents) is a real server failure → 500 without leaking internals.
fn scene_error_response(e: GroveError) -> (StatusCode, String) {
    match e {
        GroveError::Storage(msg) => (StatusCode::BAD_REQUEST, msg),
        other => {
            eprintln!("[sketches] server-side scene error: {other}");
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                "internal server error".to_string(),
            )
        }
    }
}

/// Project lookup helper that attaches a default body message so 404/5xx
/// responses carry something more useful than an empty payload. Without this
/// the frontend's `extractErrorPayload` falls back to the bare status line.
fn project_err(status: StatusCode) -> (StatusCode, String) {
    (
        status,
        status
            .canonical_reason()
            .unwrap_or("request failed")
            .to_string(),
    )
}

#[derive(Debug, Serialize)]
pub struct SketchListResponse {
    pub sketches: Vec<sketches::SketchMeta>,
}

#[derive(Debug, Deserialize)]
pub struct CreateSketchRequest {
    pub name: String,
}

#[derive(Debug, Deserialize)]
pub struct RenameSketchRequest {
    pub name: String,
}

fn internal<E: std::fmt::Display>(_e: E) -> StatusCode {
    StatusCode::INTERNAL_SERVER_ERROR
}

pub async fn list_sketches(
    Path((id, task_id)): Path<(String, String)>,
) -> Result<Json<SketchListResponse>, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    let index = sketches::load_index(&project_key, &task_id).map_err(internal)?;
    Ok(Json(SketchListResponse {
        sketches: index.sketches,
    }))
}

pub async fn create_sketch(
    Path((id, task_id)): Path<(String, String)>,
    Json(req): Json<CreateSketchRequest>,
) -> Result<Json<sketches::SketchMeta>, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    let meta = sketches::create_sketch(&project_key, &task_id, &req.name).map_err(internal)?;
    broadcast_sketch_event(SketchEvent::IndexChanged {
        project: project_key,
        task_id,
    });
    Ok(Json(meta))
}

pub async fn delete_sketch(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
) -> Result<StatusCode, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    sketches::delete_sketch(&project_key, &task_id, &sketch_id).map_err(internal)?;
    broadcast_sketch_event(SketchEvent::IndexChanged {
        project: project_key,
        task_id,
    });
    Ok(StatusCode::NO_CONTENT)
}

pub async fn rename_sketch(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
    Json(req): Json<RenameSketchRequest>,
) -> Result<StatusCode, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    sketches::rename_sketch(&project_key, &task_id, &sketch_id, &req.name).map_err(internal)?;
    broadcast_sketch_event(SketchEvent::IndexChanged {
        project: project_key,
        task_id,
    });
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, Deserialize)]
pub struct UpdateSceneRequest {
    /// Full Excalidraw scene JSON. Must be a JSON object.
    pub scene: serde_json::Value,
}

pub async fn get_scene(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
) -> Result<Response, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    let body = sketches::load_scene(&project_key, &task_id, &sketch_id)
        .map_err(|_| StatusCode::NOT_FOUND)?;
    Ok((
        [(axum::http::header::CONTENT_TYPE, "application/json")],
        body,
    )
        .into_response())
}

pub async fn put_scene(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
    Json(req): Json<UpdateSceneRequest>,
) -> Result<StatusCode, (StatusCode, String)> {
    let (_project, project_key) = find_project_by_id(&id).map_err(project_err)?;
    // Require an object scene with an `elements` array. Arbitrary JSON values
    // (null / numbers / bare arrays) would pass through raw and blow up on the
    // next read.
    if !req.scene.is_object()
        || req
            .scene
            .get("elements")
            .and_then(|v| v.as_array())
            .is_none()
    {
        return Err((
            StatusCode::BAD_REQUEST,
            "scene must be a JSON object with an `elements` array".to_string(),
        ));
    }
    let body =
        serde_json::to_string(&req.scene).map_err(|e| scene_error_response(GroveError::from(e)))?;
    sketches::save_scene(&project_key, &task_id, &sketch_id, &body)
        .map_err(scene_error_response)?;
    broadcast_sketch_event(SketchEvent::SketchUpdated {
        project: project_key,
        task_id,
        sketch_id,
        source: SketchEventSource::User,
        scene: req.scene,
    });
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, Deserialize, Default)]
pub struct PatchSceneRequest {
    #[serde(default)]
    pub created: Vec<serde_json::Value>,
    #[serde(default)]
    pub updated: serde_json::Map<String, serde_json::Value>,
    #[serde(default)]
    pub deleted: Vec<String>,
}

pub async fn patch_scene(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
    Json(req): Json<PatchSceneRequest>,
) -> Result<StatusCode, (StatusCode, String)> {
    let (_project, project_key) = find_project_by_id(&id).map_err(project_err)?;
    let body = sketches::apply_element_patch(
        &project_key,
        &task_id,
        &sketch_id,
        &req.created,
        &req.updated,
        &req.deleted,
    )
    // Client-side validation errors (missing id, duplicate id, non-object
    // update, id collision, empty-string delete) are `GroveError::Storage` →
    // 400 with message. Real server failures (I/O, on-disk JSON corruption)
    // stay 500.
    .map_err(scene_error_response)?;
    let scene: serde_json::Value =
        serde_json::from_str(&body).map_err(|e| scene_error_response(GroveError::from(e)))?;
    broadcast_sketch_event(SketchEvent::SketchUpdated {
        project: project_key,
        task_id,
        sketch_id,
        source: SketchEventSource::User,
        scene,
    });
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, Serialize)]
pub struct SketchHistoryEntry {
    pub id: String,
    pub ts: String,
    pub element_count: Option<usize>,
    pub label: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SketchHistoryResponse {
    pub entries: Vec<SketchHistoryEntry>,
}

pub async fn list_sketch_history(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
) -> Result<Json<SketchHistoryResponse>, (StatusCode, String)> {
    let (_project, project_key) = find_project_by_id(&id).map_err(project_err)?;
    let entries =
        crate::storage::sketch_checkpoints::list_entries(&project_key, &task_id, &sketch_id)
            .map_err(scene_error_response)?
            .into_iter()
            .map(|e| SketchHistoryEntry {
                id: e.id,
                ts: e.ts,
                element_count: e.element_count,
                label: e.label,
            })
            .collect();
    Ok(Json(SketchHistoryResponse { entries }))
}

#[derive(Debug, Deserialize)]
pub struct RestoreCheckpointRequest {
    pub checkpoint_id: String,
}

pub async fn restore_sketch_checkpoint(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
    Json(req): Json<RestoreCheckpointRequest>,
) -> Result<StatusCode, (StatusCode, String)> {
    let (_project, project_key) = find_project_by_id(&id).map_err(project_err)?;
    let scene = crate::storage::sketch_checkpoints::load(
        &project_key,
        &task_id,
        &sketch_id,
        &req.checkpoint_id,
    )
    .map_err(scene_error_response)?
    .ok_or_else(|| {
        (
            StatusCode::NOT_FOUND,
            format!("checkpoint '{}' not found", req.checkpoint_id),
        )
    })?;
    if !scene.is_object() || scene.get("elements").and_then(|v| v.as_array()).is_none() {
        return Err((
            StatusCode::BAD_REQUEST,
            "checkpoint payload is not a valid scene (missing `elements` array)".to_string(),
        ));
    }
    let body =
        serde_json::to_string(&scene).map_err(|e| scene_error_response(GroveError::from(e)))?;
    sketches::save_scene(&project_key, &task_id, &sketch_id, &body)
        .map_err(scene_error_response)?;
    // Broadcast as an agent-sourced update so connected clients hard-reload
    // the scene (matching AI-write behavior).
    broadcast_sketch_event(SketchEvent::SketchUpdated {
        project: project_key,
        task_id,
        sketch_id,
        source: SketchEventSource::Agent,
        scene,
    });
    Ok(StatusCode::NO_CONTENT)
}

/// PNG file signature (first 8 bytes). Rejecting non-PNG uploads prevents a
/// compromised client from storing arbitrary bytes that MCP would later hand
/// to the model as `image/png`.
const PNG_SIGNATURE: [u8; 8] = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A];

/// Accept a PNG thumbnail rendered by the web client. Staleness is detected
/// at MCP-read time via file mtime comparison (see `load_thumbnail_if_fresh`),
/// so we just write unconditionally here.
pub async fn upload_sketch_thumbnail(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
    body: Bytes,
) -> Result<StatusCode, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    if body.is_empty() {
        return Err(StatusCode::BAD_REQUEST);
    }
    if body.len() < PNG_SIGNATURE.len() || body[..PNG_SIGNATURE.len()] != PNG_SIGNATURE {
        return Err(StatusCode::BAD_REQUEST);
    }
    sketches::save_thumbnail(&project_key, &task_id, &sketch_id, &body).map_err(internal)?;
    Ok(StatusCode::NO_CONTENT)
}

/// Serve the sketch's PNG render. Returns 404 when no thumbnail has been
/// generated yet (the canvas only uploads on edit, debounced 2s); callers
/// (markdown preview) treat 404 as "fall back to plain text". Slightly stale
/// thumbs are served as-is — visual preview tolerates the lag.
pub async fn get_sketch_thumbnail(
    Path((id, task_id, sketch_id)): Path<(String, String, String)>,
    headers: HeaderMap,
) -> Result<Response, StatusCode> {
    let (_project, project_key) = find_project_by_id(&id)?;
    let Some((bytes, mtime)) =
        sketches::load_thumbnail(&project_key, &task_id, &sketch_id).map_err(internal)?
    else {
        return Err(StatusCode::NOT_FOUND);
    };
    // Weak ETag from sketch_id + mtime epoch + length so the browser can
    // 304 between markdown re-renders. Including `sketch_id` prevents the
    // (admittedly unlikely) ETag collision where two sketches happen to
    // share both byte length and mtime second. We propagate mtime errors
    // as 500 instead of silently falling back to UNIX_EPOCH — a clock
    // reading "0" would make every render share an ETag.
    let mtime_secs = mtime
        .duration_since(std::time::SystemTime::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .map_err(internal)?;
    let etag = format!("W/\"{}-{}-{}\"", sketch_id, mtime_secs, bytes.len());
    if let Some(req_etag) = headers
        .get(header::IF_NONE_MATCH)
        .and_then(|v| v.to_str().ok())
    {
        if req_etag == etag {
            return Ok(StatusCode::NOT_MODIFIED.into_response());
        }
    }
    let mut resp = (StatusCode::OK, bytes).into_response();
    let h = resp.headers_mut();
    h.insert(header::CONTENT_TYPE, HeaderValue::from_static("image/png"));
    if let Ok(v) = HeaderValue::from_str(&etag) {
        h.insert(header::ETAG, v);
    }
    h.insert(
        header::CACHE_CONTROL,
        HeaderValue::from_static("private, max-age=0, must-revalidate"),
    );
    Ok(resp)
}
